### ISSUE TITLE
Pod to pod packets via egress + ingress proxy are MTU dropped when IPsec is enabled

### ISSUE THREAD

**Issue by jschwinger233** [cilium/cilium#33168]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### What happened?

Steps:

1. Kind install cilium

```
export IMAGE=kindest/node:v1.29.4@sha256:3abb816a5b1061fb15c6e9e60856ec40d56b7b52bcea5f5f1350bc6e2320b6f8
     ./contrib/scripts/kind.sh --xdp --secondary-network "" 3 "" "" none dual 0.0.0.0 6443
 kubectl patch node kind-worker3 --type=json -p='[{"op":"add","path":"/metadata/labels/cilium.io~1no-schedule","value":"true"}]'

  if [[ "gcm(aes)" == "gcm(aes)" ]]; then
    key="rfc4106(gcm(aes)) $(dd if=/dev/urandom count=20 bs=1 2> /dev/null | xxd -p -c 64) 128"
  elif [[ "gcm(aes)" == "cbc(aes)" ]]; then
    key="hmac(sha256) $(dd if=/dev/urandom count=32 bs=1 2> /dev/null| xxd -p -c 64) cbc(aes) $(dd if=/dev/urandom count=32 bs=1 2> /dev/null| xxd -p -c 64)"
  else
    echo "Invalid key type"; exit 1
  fi
  kubectl create -n kube-system secret generic cilium-ipsec-keys \
    --from-literal=keys="3+ ${key}"
./cilium-cli install --wait     --chart-directory=./install/kubernetes/cilium     --helm-set=debug.enabled=true     --helm-set=debug.verbose=envoy     --helm-set=hubble.eventBufferCapacity=65535     --helm-set=bpf.monitorAggregation=none     --helm-set=cluster.name=default     --helm-set=authentication.mutual.spire.enabled=false     --nodes-without-cilium     --helm-set-string=kubeProxyReplacement=true     --set='' --helm-set=image.repository=quay.io/cilium/cilium-ci     --helm-set=image.useDigest=false     --helm-set=image.tag=412a46c753eaeff229cba3d83332e2dea3e192f3     --helm-set=operator.image.repository=quay.io/cilium/operator     --helm-set=operator.image.suffix=-ci     --helm-set=operator.image.tag=412a46c753eaeff229cba3d83332e2dea3e192f3     --helm-set=operator.image.useDigest=false     --helm-set=hubble.relay.image.repository=quay.io/cilium/hubble-relay-ci     --helm-set=hubble.relay.image.tag=412a46c753eaeff229cba3d83332e2dea3e192f3     --helm-set=hubble.relay.image.useDigest=false --helm-set-string=routingMode=native --helm-set-string=autoDirectNodeRoutes=true --helm-set-string=ipv4NativeRoutingCIDR=10.244.0.0/16 --helm-set-string=ipv6NativeRoutingCIDR=fd00:10:244::/56   --helm-set-string=endpointRoutes.enabled=true --helm-set=ipv6.enabled=true --helm-set=bpf.masquerade=true  --helm-set=encryption.enabled=true --helm-set=encryption.type=ipsec --helm-set=encryption.nodeEncryption=false
```

2. Install cilium-cli-next
```
 cid=$(docker create quay.io/cilium/cilium-cli-ci:5401ce3551cc46052489b7153468b577830a63a4 ls)
  docker cp $cid:/usr/local/bin/cilium .//cilium-cli-next
  docker rm $cid
```

3. Run connectivity test and see failures
```
./cilium-cli-next connectivity test --include-unsafe-tests  --flush-ct --test "pod-to-pod-with-l7-policy-encryption/"  -v  -p
```

Please note it only happens when both ingress policy and egress policy are working at the same time. Once deleting either ingress/egress policy, the connectivity is back.

### Cilium Version

```
$ ./cilium-cli version
cilium-cli: v0.16.7 compiled with go1.22.2 on linux/amd64
cilium image (default): v1.15.4
cilium image (stable): v1.15.6
cilium image (running): 1.16.0-dev
```

### Kernel Version

```
$ uname -a
Linux liangzc-l-PF4RDLEQ 6.5.0-1024-oem #25-Ubuntu SMP PREEMPT_DYNAMIC Mon May 20 14:47:48 UTC 2024 x86_64 x86_64 x86_64 GNU/Linux
```

### Kubernetes Version

```
$ kubectl version
Client Version: v1.30.0
Kustomize Version: v5.0.4-0.20230601165947-6ce0bf390ce3
Server Version: v1.29.4
```

### Anything else?

This is yet another MTU issue, ipsec xfrm makes it even harder.

## Fact one: IPsec xfrm reduces MTU from 1500 to 1446.

Even if all the net ifaces on the cilium node have been set MTU 1500 (even for those in 2005 route table), xfrm makes MTU 1446 on the sneak.

This happens in the ip_forward():

```
// https://elixir.bootlin.com/linux/v6.2/source/net/ipv4/ip_forward.c#L124
int ip_forward(struct sk_buff *skb)
{
[...]
	if (!xfrm4_route_forward(skb)) {
		SKB_DR_SET(reason, XFRM_POLICY);
		goto drop;
	}
	rt = skb_rtable(skb);

	if (opt->is_strictroute && rt->rt_uses_gateway)
		goto sr_failed;

	IPCB(skb)->flags |= IPSKB_FORWARDED;
	mtu = ip_dst_mtu_maybe_forward(&rt->dst, true);
	if (ip_exceeds_mtu(skb, mtu)) {
[...]
}
```

We know that before entering `ip_forward()` in `ip_route_input_noref()` , the skb has been set `skb->_skb_refdst`, which is used to determine MTU. However, the code above indicates the existence of xfrm could change the MTU inside `ip_forward()`: `xfrm4_route_forward()` can change `skb->_skb_refdst` to the result of `xfrm_lookup_with_ifid()`. 

The xfrm MTU has a special algorithm at https://elixir.bootlin.com/linux/v6.2/source/net/xfrm/xfrm_state.c#L2747, I didn't check all the details but I did use bpftrace to fetch the new MTU from `xfrm4_route_forward()`, in our cilium case the MTU indeed is reduced from 1500 to 1446.

## Fact two: traffic from (local egress) proxy to (remote ingress) proxy has a different MTU from pod to pod 

Either pod to pod or pod to remote ingress proxy has MTU 1423, which is set on the route instead of iface:

```
$ nspod client2-ccd7b8bdf-nt4nn ip r
default via 10.244.1.12 dev eth0 mtu 1423 
10.244.1.12 dev eth0 scope link
```

However, proxy to proxy traffic has MTU 1500. This subtle change explains why we see issues only if ingress and egress policy are installed together, the scenario we never covered in test before.

## Proposed solution

It seems we can just change MTU for routing in the 2005 table, because proxy traffic will always end up there due to 0xa00/0xb00 mark.

For example, the current routes in 2005 table are

```
$ nscontainer kind-control-plane ip r s t 2005
default via 10.244.0.100 dev cilium_host proto kernel 
10.244.0.100 dev cilium_host proto kernel scope link 
```

We could change that to

```
default via 10.244.0.100 dev cilium_host proto kernel mtu 1446
10.244.0.100 dev cilium_host proto kernel scope link mtu 1446
```

(Haven't checked IPv6)

### Cilium Users Document

- [ ] Are you a user of Cilium? Please add yourself to the [Users doc](https://github.com/cilium/cilium/blob/main/USERS.md)

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by jschwinger233** [comment]
Update: IPv6 faces the same MTU issue: xfrm reduces MTU from 1500 to 1426. But merely changing the route MTU still doesn't recover the v6 connectivity, there must be something else. So I added label `feature/ipv6` and `needs/triage`.

**Comment by whitingj** [comment]
Is there a work-around for this issue? It looks like we might be running into this.

**Comment by sgargan** [comment]
looking at [this](https://github.com/cilium/cilium/blob/main/pkg/mtu/mtu.go#L188), it seems the route mtu is always going to be less than the device mtu when encryption is enabled. In most places that I can see, the devices are configured with `GetDeviceMTU()` but probably should be configured with the min of `GetDeviceMTU` and `GetRouteMTU()`. Is there a reason to keep the Device MTU higher than the Route one?

**Comment by smagnani96** [comment]
I can get assigned to this.

I was able to reproduce this (many thanks jschwinger233 :pray: ).
For connectivity tests to run, I had to run the following commands before:
```bash
echo 1048576 | sudo tee /proc/sys/fs/inotify/max_user_watches
echo 512 | sudo tee /proc/sys/fs/inotify/max_user_instances
```

The IPv4 workaround consists into changing manually the MTU in the cilium rule table 2005 in each node (kind-worker3 doesn't need this since noSchedule).

1. List all the nodes:

```bash
CONTAINER ID   IMAGE                  COMMAND                  CREATED          STATUS          PORTS                                                                          NAMES
8e1613f9707a   kindest/node:v1.29.4   "/usr/local/bin/entr…"   17 minutes ago   Up 17 minutes   127.0.0.1:23413->2345/tcp, 127.0.0.1:23513->2346/tcp                           kind-worker3
fb1076be17e0   kindest/node:v1.29.4   "/usr/local/bin/entr…"   17 minutes ago   Up 17 minutes   127.0.0.1:23412->2345/tcp, 127.0.0.1:23512->2346/tcp                           kind-worker2
1a0dcb9517d3   kindest/node:v1.29.4   "/usr/local/bin/entr…"   17 minutes ago   Up 17 minutes   127.0.0.1:23411->2345/tcp, 127.0.0.1:23511->2346/tcp                           kind-worker
860059b63fc9   kindest/node:v1.29.4   "/usr/local/bin/entr…"   17 minutes ago   Up 17 minutes   0.0.0.0:6443->6443/tcp, 127.0.0.1:23401->2345/tcp, 127.0.0.1:23501->2346/tcp   kind-control-plane
```

2. Then connect to each node (`docker exec -ti <container_id> bash`) and run:

```bash
root@kind-control-plane:/ ip r s t 2005
default via 10.244.0.202 dev cilium_host proto kernel
10.244.0.202 dev cilium_host proto kernel scope link

root@kind-control-plane:/ ip r r t 2005 default via 10.244.0.202 dev cilium_host proto kernel mtu 1446
```

Run connectivity tests again, and only the curl-ipv6 should fail.

I'll update this issue as soon as I've updated this fix in Cilium and investigated the IPv6 behavior.

**Comment by smagnani96** [comment]
I'm back at this issue with some good updates for IPv6.

Considering the minimum network packet with the total size of **48B** (original header+original payload), those should be the differences between v4 and v6 (only in the new IPSec header).

|  | Bytes (v4) | Bytes (v6) |
| --- | --- | --- |
| New IPSec Header | 20 | 40 |
| ESP Header (spi+sequence) | 8 | 8 |
| ESP IV (implicit initialization vector) | 16 | 16 |
| Original Header | 20 | 40 |
| Original Payload | 28 | 8 |
| ESP Trailer (pad, pad length, next header) | 16 | 16 |
| --- | --- | --- |
| **Total** | **108** | **128** |

The solution for restoring connectivity in v4 was setting the MTU to 1446. Therefore, I replicated the same for the v6 routing table with MTU = 1446-20 = 1426 and it worked (MTUs >1426 break connectivity).

Commands are as follow (to be run on both the two nodes):

```bash
root@kind-control-plane:/# ip -6 r s t 2005
fe80::f495:98ff:fe95:b18a dev cilium_host proto kernel metric 1024 pref medium
default via fe80::f495:98ff:fe95:b18a dev cilium_host proto kernel metric 1024  pref medium

root@kind-control-plane:/# ip -6 r r t 2005 default via fe80::f495:98ff:fe95:b18a dev cilium_host proto kernel metric 1024  pref medium mtu 1426
```

To wrap up, I think it makes sense to reduce the MTU for both IPv4 and IPv6 when the egress+ingress proxy and IPSec are enabled simultaneously.

**Comment by smagnani96** [comment]
I'm back at this issue with some updates for IPv6.

## IPv6 investigation

Considering the minimum network packet with the total size of **48B** (original header+original payload), those should be the differences between v4 and v6 (only in the new IPSec header), as also @jschwinger233 mentioned.

  | Bytes (v4) | Bytes (v6)
-- | -- | --
New IPSec Header | 20 | 40
ESP Header (spi+sequence) | 8 | 8
ESP IV (implicit initialization vector) | 16 | 16
Original Header | 20 | 40
Original Payload | 28 | 8
ESP Trailer (pad, pad length, next header) | 16 | 16
--- | --- | ---
Total | 108 | 128


> Update: IPv6 faces the same MTU issue: xfrm reduces MTU from 1500 to 1426. But merely changing the route MTU still doesn't recover the v6 connectivity, there must be something else. So I added label feature/ipv6 and needs/triage.

The solution for restoring connectivity in v4 was setting the MTU to 1446. Therefore, I replicated the same for the v6 routing table with MTU = 1446-20 = 1426 and it worked (MTUs >1426 break connectivity), differently than what mentioned in the quoted comment.

Commands are as follow (to be run on both the two nodes):

```bash
root@kind-control-plane:/# ip -6 r s t 2005
fe80::f495:98ff:fe95:b18a dev cilium_host proto kernel metric 1024 pref medium
default via fe80::f495:98ff:fe95:b18a dev cilium_host proto kernel metric 1024  pref medium

root@kind-control-plane:/# ip -6 r r t 2005 fe80::f495:98ff:fe95:b18a dev cilium_host proto kernel metric 1024 pref medium mtu 1426
root@kind-control-plane:/# ip -6 r r t 2005 default via fe80::f495:98ff:fe95:b18a dev cilium_host proto kernel metric 1024  pref medium mtu 1426

root@kind-control-plane:/# ip -6 r s t 2005
fe80::f495:98ff:fe95:b18a dev cilium_host proto kernel metric 1024 pref medium mtu 1426
default via fe80::f495:98ff:fe95:b18a dev cilium_host proto kernel metric 1024  pref medium mtu 1426
```

## Concerns

After a quick chat with @dylandreimerink the following doubt came up.

> From the MTU perspective we have 2 situations. Creation of a packet and forwarding of a packet. If we install a route in a routing table which only forwards an existing packet then it will fragment the packet which we don't want. But if we set it on the route table where the packet is created (the same network namespace) then the stack of the kernel will send packets that respect the MTU.
So the question is, is the proxy using the host NS, if so the 2005 routes might work. But if the proxy sits in its own namespace and we only forward packets to the host NS after the creation, we get fragmentation which is undesirable if not breaks traffic. So we want to set the MTU as close to the source as possible

With the configuration provided, changing mtu in the routes in 2005 table seems ok and does not produce fragmentation. Do we have cases in which proxies sit in separate namespaces? (I do not have the answer yet, still investigating)

**Comment by jschwinger233** [comment]
Thank you @smagnani96 

I think Dylan was talking about a possible improvement of proxy: moving proxy into an isolated netns, to solve issues like "conflicted port in host netns" or "wrong socket lookup in host netns". 

If that happens someday in future, current mark-based route table doesn't work for proxy anymore, because it tries to route skb with mark 0xa00/0xf00 or 0xb00/0xf00 to 2005 table, but skb->mark will be reset to 0 after going across netns. We'll have to figure out another approach after all, so I'm not super concerned about setting route to 2005 route table for now.

**Comment by smagnani96** [comment]
Thanks to both @dylandreimerink and @jschwinger233  for the feedback. We'll surely have to account for that if/when moving proxy to a separate namespace. For now, it seems ok to modify the table 2005 to restore connectivity. I've edited/updated the previous investigation.

After a discussion with @jschwinger233, we could simply set the same MTU as for the pod-to-pod or pod-to-remote-ingress route, for which we already account for the IPSec overhead. This should restore both IPv4 and IPv6 connectivity.

## Future Issue

Soon, I'll open a new issue to account for the dynamic changing of this IPSec overhead, which right now we compute only once. When rotating an IPSec key with a complete new algorithm, that IPSecOverhead might change, potentially causing packet drop when increasing (should use a lower MTU). An example is when rotating a key:

* from `3+ rfc4106(gcm(aes)) $(shell dd if=/dev/urandom count=20 bs=1 2> /dev/null | xxd -p -c 64) 128`
* to `4+ hmac(sha256) 0123456789abcdef0123456789abcdef cbc(aes) 0123456789abcdef0123456789abcdef`

In that case, the `authKeySize` would change from 16 to 32, so if we don't update the MTU in the route accordingly we'll likely hit connectivity problems (v4/v6 depending of how much this field increases).
